From 3d8a4205993cc07deae970c00f2e2d83f6ffe618 Mon Sep 17 00:00:00 2001
From: Valentin Caron <valentin.caron@foss.st.com>
Date: Fri, 7 Feb 2025 12:28:04 +0100
Subject: [PATCH 1300/1695] drivers: stm32_rtc: fix overflow at midnight

This is what happens when we call this function with day switch between
cur and ref parameters:

diff_in_tick += -39188232 (subs)
diff_in_tick += 18443288113709551 (time)
diff_in_tick += 3456000000000 (date)

instead of

diff_in_tick += -39188232 (subs)
diff_in_tick += -3456000000000 (time)
diff_in_tick += 3456000000000 (date)

Fix this with UDIV_ROUND_NEAREST macro to handle correctly the division.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git b8b87c015ef21a481892b9b5bad47d40cea4ff98 drivers: stm32_rtc: fix overflow at midnight]
Signed-off-by: Valentin Caron <valentin.caron@foss.st.com>
Change-Id: I442470275319c0332a788cc3ce51272a22472ffe
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/435470
Domain-Review: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/stm32_rtc.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/core/drivers/stm32_rtc.c b/core/drivers/stm32_rtc.c
index cd8019f70..c030f7c66 100644
--- a/core/drivers/stm32_rtc.c
+++ b/core/drivers/stm32_rtc.c
@@ -528,9 +528,9 @@ unsigned long long stm32_rtc_diff_calendar_tick(struct stm32_rtc_calendar *cur,
 
 	diff_in_tick += stm32_rtc_diff_subs_tick(cur, ref, tick_rate);
 	diff_in_tick += stm32_rtc_diff_time_ms(&curr_t, &ref_t) *
-			tick_rate / MS_PER_SEC;
+			UDIV_ROUND_NEAREST(tick_rate, MS_PER_SEC);
 	diff_in_tick += stm32_rtc_diff_date_ms(&curr_t, &ref_t) *
-			tick_rate / MS_PER_SEC;
+			UDIV_ROUND_NEAREST(tick_rate, MS_PER_SEC);
 
 	if (diff_in_tick < 0)
 		panic("Negative time difference is not allowed");
-- 
2.47.2

